panther_battery/src/dual_battery_publisher.cpp

```diff
@@ -0,0 +1,173 @@
+    battery_msg.power_supply_health = BatteryStateMsg::POWER_SUPPLY_HEALTH_OVERVOLTAGE;
+    battery_msg.temperature = std::max<float>(battery_msg_1.voltage, battery_msg_2.voltage);
+  } else if (
+    battery_msg_1.power_supply_health == BatteryStateMsg::POWER_SUPPLY_HEALTH_COLD ||
+    battery_msg_2.power_supply_health == BatteryStateMsg::POWER_SUPPLY_HEALTH_COLD) {
+    battery_msg.power_supply_health = BatteryStateMsg::POWER_SUPPLY_HEALTH_COLD;
+    battery_msg.temperature = std::min<float>(battery_msg_1.temperature, battery_msg_2.temperature);
+  } else if (
+    battery_msg_1.power_supply_health == BatteryStateMsg::POWER_SUPPLY_HEALTH_UNKNOWN ||
+    battery_msg_2.power_supply_health == BatteryStateMsg::POWER_SUPPLY_HEALTH_UNKNOWN) {
+    battery_msg.power_supply_health = BatteryStateMsg::POWER_SUPPLY_HEALTH_UNKNOWN;
+  } else {
+    battery_msg.power_supply_health = BatteryStateMsg::POWER_SUPPLY_HEALTH_GOOD;
+  }
+
+  return battery_msg;
+}
+
+void DualBatteryPublisher::ValidateMergeBatteryMsgs(
+  const BatteryStateMsg & battery_msg_1, const BatteryStateMsg & battery_msg_2) const
+{
+  if (battery_msg_1.header.stamp != battery_msg_2.header.stamp) {
+    throw std::runtime_error("Message header stamp mismatch");
+  } else if (battery_msg_1.power_supply_technology != battery_msg_2.power_supply_technology) {
+    throw std::runtime_error("Battery power supply technology mismatch");
+  } else if (battery_msg_1.cell_voltage.size() != battery_msg_2.cell_voltage.size()) {
+    throw std::runtime_error("Battery cell voltage mismatch");
+  } else if (battery_msg_1.cell_temperature.size() != battery_msg_2.cell_temperature.size()) {
+    throw std::runtime_error("Battery cell temperature mismatch");
+  } else if (battery_msg_1.location != battery_msg_2.location) {
+    throw std::runtime_error("Battery location mismatch");
+  } else if (battery_msg_1.present != battery_msg_2.present) {
+    throw std::runtime_error("Battery present mismatch");
+  }
+}
+
+uint8_t DualBatteryPublisher::MergeBatteryPowerSupplyStatus(
+  const BatteryStateMsg & battery_msg_1, const BatteryStateMsg & battery_msg_2) const
+{
+  if (battery_msg_1.power_supply_status == battery_msg_2.power_supply_status) {
+    return battery_msg_1.power_supply_status;
+  } else if (
+    battery_msg_1.power_supply_status == BatteryStateMsg::POWER_SUPPLY_STATUS_UNKNOWN ||
+    battery_msg_2.power_supply_status == BatteryStateMsg::POWER_SUPPLY_STATUS_UNKNOWN) {
+    return BatteryStateMsg::POWER_SUPPLY_STATUS_UNKNOWN;
+  } else if (
+    battery_msg_1.power_supply_status == BatteryStateMsg::POWER_SUPPLY_STATUS_DISCHARGING ||
+    battery_msg_2.power_supply_status == BatteryStateMsg::POWER_SUPPLY_STATUS_DISCHARGING) {
+    throw std::runtime_error(
+      "Detected critical mismatch between battery 1 and battery 2 messages. Battery 1 indicates "
+      "discharging, while battery 2 indicates charging state.");
```

Comment:
Again this is more safe thinking. If the way charging is detected changes this will ensure that underlying logic for this method don't have to be changed

Reply:
If we want to leave it, I think that order should be changed (`POWER_SUPPLY_STATUS_NOT_CHARGING` before `POWER_SUPPLY_STATUS_DISCHARGING`) and handling `POWER_SUPPLY_STATUS_FULL` should be added